src/parser/jobs/sam/modules/Sen.tsx

```diff
@@ -1,98 +1,307 @@
+							lastSenState.isOverwrite = true
+						}
+
+						break
+				}
+			}
+		}
+	}
+
+// Function that handles SenState check, if no senState call the maker and then push to the rotation
+	private checkCastAndPush(event: CastEvent) {
+
+		// step 1: set action
+		const action = event.ability.guid
+
+		// step 2: FILTER EVERYTHING
+
+		if (THINGS_WE_WANT_IN_THE_TABLE.includes(action) || SEN_ACTIONS.includes(action) || SEN_REMOVERS.includes(action) ) {
+
+			// step 3: check the sen state, if undefined/not active, make one
+
+			let lastSenState = this.lastSenState
+
+			if ((typeof lastSenState === 'undefined') || (lastSenState.isDone === true) ) {
+
+				this.senStateMaker(event)
+			}
+
+			lastSenState = this.lastSenState
+
+			if (lastSenState != null && lastSenState.end == null) { // The state already exists
+
+				// Push action
+				lastSenState.rotation.push(event)
+			}
+
+		}
+
+	}
+
+// Make a new sen state!
+	private senStateMaker(event: CastEvent) {
+		const senState = new SenState(event.timestamp)
+		this.senStateWindows.push(senState)
+
+	}
+
+	private senCodeProcess() {
+		const lastSenState = this.lastSenState
+
+		if (lastSenState != null && lastSenState.end == null) {
+			// Drop down the totem pole
 
-		if (this.sen[sen]) {
-			this.wasted++
+			if (lastSenState.isDeath === true) {
+				lastSenState._senCode = SEN_HANDLING.DEATH
+				lastSenState.isNonStandard = true
+			}
+			else if (lastSenState.isOverwrite === true) {
+				lastSenState._senCode = SEN_HANDLING.OVERWROTE_SEN
+				lastSenState.isNonStandard = true
+			}
+			else if (lastSenState.isHaga === true) {
+				lastSenState._senCode = SEN_HANDLING.HAGAKURE
+				lastSenState.isNonStandard = true
+			}
 		}
 
-		this.sen[sen] = true
 	}
 
-	private remove() {
-		this.sen = _.mapValues(this.sen, () => false)
+// End the state, count wastes, add it
+	private remove(event: CastEvent) {
+		const lastSenState = this.lastSenState
+
+		if (lastSenState != null && lastSenState.end == null) {
+
+			if (event.ability.guid === ACTIONS.HAGAKURE.id) {
+				lastSenState.kenkiGained = (lastSenState.currentSetsu + lastSenState.currentGetsu + lastSenState.currentKa) * KENKI_PER_SEN
+				lastSenState.isHaga = true
+
+				this.kenki.modify(lastSenState.kenkiGained)
+			}
+
+			this.wasted = this.wasted + (lastSenState.overwriteSetsus + lastSenState.overwriteGetsus + lastSenState.overwriteKas)
+
+			lastSenState.isDone = true
+			this.senCodeProcess()
+			lastSenState.end = event.timestamp
+		}
 	}
 
-	private onHagakure() {
-	// work out how many sen are currently active
-	const activeSen = Object.values(this.sen)
-		.filter(active => active)
-		.length
+	private onDeath() {
+		const lastSenState = this.lastSenState
+
+		if (lastSenState != null && lastSenState.end == null) {
+			this.wasted = this.wasted + (lastSenState.overwriteSetsus + lastSenState.overwriteGetsus + lastSenState.overwriteKas) + (lastSenState.currentSetsu + lastSenState.currentGetsu + lastSenState.currentKa)
```

Comment:
Even if there were no Sen lost due to the death, this is still marking the Sen window as non-standard due to a death and adding it to the output table (see http://localhost:3000/analyse/dvhD12atwZFLz8AN/3/11/).  Should those cases be excluded from the output table?